File: interviewbook/stackandqueue/getMinStack/getMinStack.go
package main

import "fmt"

// Solution1:
// Stack implement stack
// type Stack []interface{}

// func (s Stack) isEmpty() bool {
// 	return len(s) <= 0
// }

// func (s *Stack) push(val interface{}) {
// 	stack := *s
// 	if s.isEmpty() {
// 		*s = append(*s, val)
// 		*s = append(*s, val)
// 		return
// 	}
// 	curMin := stack[len(stack)-1]
// 	if curMin.(int) >= val.(int) {
// 		curMin = val
// 	}
// 	*s = append(*s, val)
// 	*s = append(*s, curMin)
// }

// func (s *Stack) pop() (interface{}, error) {
// 	stack := *s
// 	if stack.isEmpty() {
// 		return nil, errors.New("empty")
// 	}
// 	val := stack[len(stack)-2]
// 	*s = stack[:len(stack)-2]
// 	return val, nil
// }

// func (s Stack) peek() (interface{}, error) {
// 	if s.isEmpty() {
// 		return nil, errors.New("empty")
// 	}
// 	return s[len(s)-2], nil
// }

// func (s Stack) getMin() (interface{}, error) {
// 	if s.isEmpty() {
// 		return nil, errors.New("empty")
// 	}
// 	return s[len(s)-1], nil
// }

// func main() {
// 	var stack Stack
// 	for i := 0; i < 5; i++ {
// 		stack.push(i)
// 	}

// 	for i := -1; i > -5; i-- {
// 		stack.push(i)
// 	}
// 	curMin, _ := stack.getMin()
// 	curTop, _ := stack.pop()

// 	curMin1, _ := stack.getMin()
// 	curTop1, _ := stack.peek()

// 	fmt.Printf("min: %d, top: %d, \nmin: %d, top: %d", curMin, curTop, curMin1, curTop1)
// }

// Solution2:

// Node implement listnode
type Node struct {
	val  int
	min  int
	next *Node
}

// List implememt linkedlist
type List struct {
	node *Node
	len  int
}

// New implement init
func New() *List {
	return &List{
		node: nil,
		len:  0,
	}
}

func (l List) isEmpty() bool {
	return l.len <= 0
}

func (l *List) push(val int) {
	n := l.node
	if n == nil {
		n = &Node{
			val:  val,
			min:  val,
			next: nil,
		}
		l.node = n
		l.len++
		return
	}
	curMin := n.min
	if curMin >= val {
		curMin = val
	}
	newHead := &Node{
		val:  val,
		min:  curMin,
		next: n,
	}
	l.node = newHead
	l.len++
}

func (l *List) pop() {
	if l.isEmpty() {
		return
	}
	n := l.node
	temp := n.next
	n.next = nil
	l.node = temp
	l.len--
}

func (l *List) top() int {
	if l.isEmpty() {
		return -1
	}
	n := l.node
	return n.val
}

func (l *List) getMin() int {
	if l.isEmpty() {
		return -1
	}
	n := l.node
	return n.min
}

func main() {
	list := New()
	for i := 0; i < 5; i++ {
		list.push(i)
	}

	for i := -2; i > -7; i-- {
		list.push(i)
	}

	curMin := list.getMin()
	curTop := list.top()

	list.pop()

	curMin1 := list.getMin()
	curTop1 := list.top()
	fmt.Printf("min: %d, top: %d, \nmin: %d, top: %d", curMin, curTop, curMin1, curTop1)
}

File: interviewbook/stackandqueue/getMinStack/getMinStack.java
package interviewbook.stackandqueue.getMinStack;
import java.util.*;
// Solution1:
// class getMinStack {
//     private Deque<Integer> norStack;
//     private Deque<Integer> minStack;

//     public getMinStack() {
//         norStack = new ArrayDeque<>();
//         minStack = new ArrayDeque<>();
//     }

//     private int pop() {
//         if (norStack.isEmpty()) {
//             throw new EmptyStackException();
//         }
//         int val = norStack.pollFirst();
//         minStack.pollFirst();
//         return val;
//     }

//     private void push(int val) {
//         norStack.offerFirst(val);
//         if (minStack.isEmpty()) {
//             minStack.offerFirst(val);
//             return;
//         }
//         int curMin = minStack.peekFirst();
//         curMin = curMin < val ? curMin : val;
//         minStack.offerFirst(curMin);
//     }

//     private int getMin() {
//         return minStack.peekFirst();
//     }

//     public static void main(String[] args) {
//         getMinStack stack = new getMinStack();
//         for (int i = 0; i < 5; i++) {
//             stack.push(i);
//         }
//         for (int i = -5; i > -10; i--) {
//             stack.push(i);
//         }
//         int curMin = stack.getMin();
//         int top = stack.pop();
//         int curMin1 = stack.getMin();
//         System.out.println("min: " + curMin + "top: " + top + "min: " + curMin1);
//     }
// }

// Solution2:
class getMinStack {
    class Node {
        Node next;
        int val;
        int min;
        public Node(int val, int min) {
            this.val = val;
            this.min = min;
            this.next = null;
        }
    }

    Node head;
    public getMinStack() {
        this.head = null;
    }

    public void push(int val) {
        if (this.head == null) {
            this.head = new Node(val, val);
            return;
        }
        
        int curMin = head.min < val ? head.min : val;
        Node newHead = new Node(val, curMin);
        newHead.next = this.head;
        this.head = newHead;
        return;
    }

    public void pop() {
        if (this.head == null) {
            throw new EmptyStackException();
        }
        Node newHead = head.next;
        head.next = null;
        head = newHead;
    }

    public int top() {
        if (this.head == null) {
            throw new EmptyStackException();
        }
        return head.val;
    }

    public int getMin() {
        if (this.head == null) {
            throw new EmptyStackException();
        }
        return head.min;
    }

    public static void main(String[] args) {
        getMinStack stack = new getMinStack();
        for (int i = 0; i < 5; i++) {
            stack.push(i);
        }
        for (int i = -1; i > -5; i--) {
            stack.push(i);
        }
        System.out.println("min: " + stack.getMin() + "top: " + stack.top());
        stack.pop();
        System.out.println("min: " + stack.getMin() + "top: " + stack.top());
    }
}